### ISSUE TITLE
Kafka external stream support csv etc input formats

### ISSUE THREAD

**Issue by chenziliang** [timeplus-io/proton#175]
**Use case**

Support input formats here : https://clickhouse.com/docs/en/interfaces/formats

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by zliang-min** [comment]
Let's firstly clarify what does it mean to support CSV for kafka external stream. I suppose how this works is that each message read from the kafka topic is a line of CSV, and the number of values matches the number of fields of the stream. For example, given a kafka external stream:
```sql
CREATE EXTERNAL STREAM employees (
  id string,
  name string,
  job_title string,
  joined_on datetime64,
  level integer
) SETTINGS type='kafka'
```

It expects that each kafka message looks like this:
```csv
1001,Some Body,Software Engineer,2023-10-01T09:00:00Z,5
```

Also, we will need to convert the values to the types specified in the stream schema. That means, in the above case, it needs to convert the `joined_on` value to type `datetime64`, and `level` to `integer`.

Does this make sense?

**Comment by chenziliang** [comment]
Yes, a bit more detail https://clickhouse.com/docs/en/engines/table-engines/integrations/kafka and we will need figure out how https://clickhouse.com/docs/en/interfaces/formats works.

But in general, each format expects a specific form of input and do parsing, conversion etc on the fly to match the defined schema.

**Comment by chenziliang** [comment]
And this integration shall be very nature and the formats code just does the work and we will be careful about perf and need documents the input details with examples.

**Comment by chenziliang** [comment]
Protobuf has highest priority.

**Comment by zliang-min** [comment]
Reading from kafka external streams with CSV format actually already works today, and in fact, all supported input data format should work too ( but I didn't test all of them, I just tested CSV and JSONEachRow so far ). However, there is a major caveat ( or bug ): the external stream cannot user `_tp_time` ( probably, all T+ internal fields can't be used, I tried `_tp_sn` which will fail as well ). For example, given stream:
```sql
CREATE EXTERNAL STREAM foo (
  a string,
  b int64
) SETTINGS type='kafka',data_format='CSV',...
```
and the target Kafka topic contains messages like `some string, 10`, `select * from foo` works as expected. However, if we create a stream like
```sql
CREATE EXTERNAL STREAM foo (
  _tp_time datetime64,
  a string,
  b int64
) SETTINGS type='kafka',data_format='CSV',...
```

`select * from foo` won't work no matter how the messages look like in the Kafka topic. The reason is that, when the formatter calculates the columns ( in `DB::IRowInputFormat::generate` ), it only counts `a` and `b`. However, when it tries to read the field (in `DB::RowInputFormatWithNamesAndTypes::readRow`), it will count the `_tp_time` column as well. I didn't check how difficult it is to fix this, but I wonder if it's worth fixing at all. Because even with the second example, `select a, b from foo` can still work.

**Comment by chenziliang** [comment]
When i initially implemented external stream, CSV was just ad-hoc tested. There are probably quite a few caveats like error handling, performance etc (same for other types etc). We will need fix _tp_time, if in the schema, users provide a _tp_time column, we use it, otherwise use `_tp_append_time` as `_tp_time`. Probably this is the best / closest thing we can do. 

virtual column `_tp_sn`, `_tp_append_time` shall always work, otherwise bugs as well, we will need fix them.

**Comment by chenziliang** [comment]
There is another requirement regarding protobuf, users like explicitly map protobuf fields to a different column name, like in protobuf we have `product.name`, they like to use `product_name` as the column name.

**Comment by chenziliang** [comment]
_tp_time etc special columns need to be added / excluded in different places to make things happy. Worth the efforts for sure.
